### DISCUSSION TITLE
Nutritional Information per food item

### DISCUSSION THREAD

**Discussion by MarthinusBosman** [mealie-recipes/mealie#2748] [Feature Request]
### First Check

- [X] I used the GitHub search to find a similar requests and didn't find it.
- [X] Checked the [tasks tagged](https://github.com/hay-kot/mealie/issues?q=is%3Aissue+is%3Aopen+label%3Atask+) issues and verified my feature is not covered

### Please provide a concise description of the problem that would be addressed by this feature.

Currently, nutritional info needs to be set per recipe. IMO it would make a lot more sense to have this info on a food item level and then just aggregate for each recipe.

Could then be combined with existing databases like fatsecret to get the nutritional information of common food items automatically during seeding.

### Please provide a concise description of the feature that would resolve your issue.

Add nutritional info per food item

### Please consider and list out some caveats or tradeoffs made in your design decision

Might break existing recipe nutritional info, but if it's implemented in a way that nutritional info can still be set manually per recipe to override default values (in cases where some food items are unavailable in list of food items for instance) then it should be fine.

### Additional Information

- [X] If this is accepted I'm willing to submit a PR to provide this feature
- [X] If this is accepted I'm willing to help maintain this feature
- [ ] I'm willing to sponsor/pay a developer to do this work

**Comment by RealFoxie** [comment]
The main problem here is that a recipe has an `amount` and a specific `unit` for a certain food.
A recipe with "200g butter" or "200cups butter" will have different nutritional values.
You could do very straightforward aggregation (ignore amount and unit), but then it really isn't that useful, is it?

**Reply by MarthinusBosman** [reply]
I assumed nutritional info would be stored relative to some unit. Though that does mean we'd need to be able to actually convert between units, which honestly in itself is a good feature as well.

**Reply by RealFoxie** [reply]
Yes, but that's not trivial. I do this now in an external script to calculate cost of recipes, but as you can have any number of units and use them with any food, you need to add some constraints and such, which is currently not a thing.

**Reply by MarthinusBosman** [reply]
I see the backend repo is in python fastapi, I've got quite some experience with that. if I get some time I'll try and implement the whole solution.

**Reply by monksy** [reply]
Breaking out the unit, quanity and food seems like it needs some work. When I import recipes it looks like it's just copying the text there.

**Reply by Yuni1988** [reply]
I'd say that a solution where you add nutrition data per unit would still be very useful. E.g. nutrition data for 100g, nutrition data for 1 cup... And if nutrition data for specific food+unit is missing, do not include it in a calculation (just add some sort of visual indication when that happens). The user could have the ability to choose between using calculated nutrition data or overriding nutrition data for the whole meal (which would also make it backwards compatible with the current system).

If calculations between units are desired, user could enter conversion data instead of nutrition data for that unit. For example 1 cup of milk = 240g of milk. This way the nutrition data can be given per 100g, the rest can be derived from that. 

Another possibility would be to make an importer for the USDA database which could seed the tables for food & nutrition data. I saw there are recent JSON files available for the database. But personally I wouldn't mind entering the data on my own, it's a one-time effort after all.

**Reply by RealFoxie** [reply]
Yes, some conversion logic is possible. Tandoor seems to have it in some way (but I have never used it before). Still a very hard problem to implement though, so won't happen if somebody who really wants it also has the experience and time to do so.

Basically all you need is conversions between units and specific conversions per food of needed (e.g. Grams to kilograms is done in units, but cl whip cream to grams whip cream will have to be a food specific conversion).
And then some graph logic that takes the path to create the conversions.

But that explanation skips over how to make this clear for the end user, a working editor view, decision when to show the conversion and when not etc.

**Comment by Eluinn** [comment]
Specifically came looking for thoughts along these lines and I would also love more in-depth nutritional data!

One idea: Nutritional data can be pulled from https://fdc.nal.usda.gov/ there is an API that can be used as well. Might require naming ingredients more specifically ie instead of just 'flour', you may need to put 'Flour, wheat, all-purpose, enriched, unbleached' (nutritional info for this particular entry here> https://fdc.nal.usda.gov/fdc-app.html#/food-details/789951/nutrients) or maybe a drop down, or search-based list population to choose from when importing a recipe. Such as:

Ingredient listed as Flour, please select which type: Flour, 00; Flour, almond; Flour, amaranth; Flour, pastry, unenriched, unbleached; Flour, wheat, all-purpose, enriched, unbleached; etc....

From what I've seen, the data is collected by using 100 gram samples, so this might only work if the ingredient itself is listed in grams/weight, unless an ingredient conversion is built in.

Using https://fdc.nal.usda.gov/ would also enable more in depth nutritional data, such as vitamin and mineral composition as well, and it would regularly update as new ingredients are added or retested.

There are also legacy entries such as: https://fdc.nal.usda.gov/fdc-app.html#/food-details/168936/nutrients that have even more details, such as amino acids and the like, but not sure how those fit in with the API.

EDIT:
For anyone who's interested, here are some google sheet templates I made with easy scaling, costing, and nutrition back when I was in culinary school. Back then I used https://Nutritiondata.self.com but their database (which pulled from https://fdc.nal.usda.gov/) is no longer available. I even set up auto generating charts for Caloric Ratio and an Amino Acid Profile.
I would love the ability to have similar use and available data within mealie!

The sheets are based entirely on weight and ratios (baker's %), but I built in some conversion tables into the documents as well.
Green sections are where you enter your info, red does calculations (try not to mess with these), and when you enter an ingredient in the template, it parses the database sheet for all the info.

Costing and Nutrition database: https://docs.google.com/spreadsheets/d/1mr6b6tyhLIi48sHVy7gICHRgDab3vKsxC1cuOxmEM5Q/edit?usp=drive_link

Recipe Collection template: https://docs.google.com/spreadsheets/d/173_fO3BrywULO2Th_M1MOACZ-MyjNyp9yDgL-PKHQXo/edit?usp=drive_link

**Comment by VitoFe** [comment]
Well, in order to preserve sanity and keep thing simple, it could be enabled like an optional feature to manually manage a list of ingredients with associated nutritional facts on a per gram basis, since grams are internationally used in this field, so that when using such ingredient in a recipe, the nutritional facts could be auto-calculated and updated depending on serving size too. Or you could also get that info from a database like apps like MyFitnessPal do. Thing is, if I have to keep track of macros/calories and the recipe database forces me to re-calculate nutritional facts every time I change something in the recipe manually, at that point I'd rather stick to my excel sheet alone, as the mealie would just be redundant eye candy.

**Reply by Kuchenpirat** [reply]
Having just one unit would simplify things, but i feel that is a verry incomplete solution i don't think adds much because you will have to do a lot of recalculation regardless. 
Having anything in grams seems not that practical to me. Like lots of vegtables and greens will be just without any units (like 1 onion, 3 tomatoes etc).

my guess is that without having a base unit for weigt and fluids and from there on relations to the other units an implementation of this feature will always be lacking and does do more to overload the app than do good. 

Sorry that we cannot meet your exact usecase here, but i don't think this makes sense at the moment.

**Comment by aredon** [comment]
Tandoor does include unit conversions and aggregates calorie and macro information from the food items. While I understand this is difficult to implement I feel it is a critical piece toward something that is missing from most meal planners:
 - A caloric and macro nutrient summary for the day.
- Generating meal plans based on macro targets

As it is if I'm trying to hit 150g of protein I'm still doing all of that by hand which defeats the point of a meal planner. At that point it's just a notepad with extra steps.

Would love to see someone create this on a self hosted service. It's present on lots of web based apps but I don't really want a cloud solution.

**Reply by Oettinger77** [reply]
Yes, this is a big big plus for Tandoor. Would like to see that in Mealie.
(I am still testing both and havn't decided yet. Tending towards Mealie, but atm Tandoor has some nice arguments).

**Comment by JurgenCruz** [comment]
I used to use this app for this tally and it handles unit conversions for you:
https://play.google.com/store/apps/details?id=com.cronometer.android.gold

![Screenshot_20250604-135614_Cronometer](https://github.com/user-attachments/assets/f87bd0c0-2d6d-4b1c-89f1-c5668f670c13)

If you change the unit drop down it will recalculate. Once you add multiple ingredients, it will tally the full recipe:

![Screenshot_20250604-135859_Cronometer](https://github.com/user-attachments/assets/3601d09f-93e8-48bc-8f6f-7adf649ebf41)

My guess is its source uses a single unit and internally it has a table for conversions between units

**Comment by EinToni** [comment]
I also think the user would greatly benefit from providing nutrition information per ingredient rather then per recipe.
I thought about how to provide this for a bit and checked out how tandoor recipes did it and from my point of view this could be done in a few nicely separated steps:

1. **Add nutrition data to ingredient**
Provide the ability to enter a [FDC](https://fdc.nal.usda.gov/food-search) ID (or theoretically some other DB) for each ingredient.  This then simply provides the nutritional values per 100g (the default) for a fixed set of nutritional values (e.g. energy, protein, fat...). (I also read ideas about atomatically searching in here, but this would require translation, etc. and would therefore be too much of a hassle imho)
_(optional)_ Alternatively the user should also be able to input custom nutritional values on their own for e.g. foods that are not in the database or are selfmade or what ever.
_(optional)_ For more customization the user could also define nutrition fields on its own, but this add quite a bit more complexity and I don't know if there is any added benefit that the user can add custom fields if the fixed set of nutriton fields are broad enough. (Tandoor let's the user specify each nutritional value on their own and then add the ID from the FDC to it for automatic pull from the DB, but imho this is too complicated).

2. **Provide the unit per ingredient**
Units are currently only strings, therefore we have to provide some relation to the "scientific" 100g. This could be done on unit level or on ingredient level. Both of this seems quite reasonable, but we shouldn't over-complicate it.
2.1 **Unit level**
This is quite easy for the user as there has just to be defined e.g. 1 tbsp equals 10g (example). But as e.g. 1 cup of flour weights differently from 1 cup of oil this is quite inaccurate. It could be used as a "fallback" default, but well... don't know how useful it is.
2.2 **Ingredient level**
On an ingredient level this is more work (for the user) but provides higher accuracy. I imagine the user to input the following for each ingredient: "\<selectable amount\> of \<selectable unit\> equals \<selectable amount\> grams". I don't think custom conversions from "any" to "any" is too helpful (like done in tandoor) as we really only need the conversion from anything to grams.
This way the user has to define how much grams 1 potato is or 1 cup of flour, but I think this would provide the best results.
_(optional)_ The FDC also provides measures for quite a few ingredients, e.g. [carrots](https://fdc.nal.usda.gov/food-details/170393/measures). This could be used to perform conversions automatically but again, a lot of things could go wrong here and we would even first translate things into english... so manual input it the way to go I would say.

3. **Sum up the nutrition**
Well this should now be really easy. Take all the ingredients of a recipe and sum up all of the nutritional values available. If there are some missing, it would be probably best to show some warning.

I did not (yet) contribute to mealie and therefore cannot say how easy things are to implement, but after using it, it looks quite well structured and I do imagine the code and database to be structure quite good as well. I would be happy to hear some feedback and also open to start working things I listed.

**Comment by bobloadmire** [comment]
wondering if there has been any progress on this feature since July?

**Comment by infectiousstupidity** [comment]
This would be a huge QoL improvement. In my opinion it doesn't have to be perfect or polished, e.g the nutritional value of an onion is not something I consider important, but being able to at least get an estimate of the total calories and macros is a big win regardless, without having to manually calculate it per recipe which is incredibly tedious. So even just something that is able to provide data about e.g rice, potatoes, legumes and meat sources in grams would be big.

But this is coming from someone who does not use vague "cups" or "a pinch of" measurements when cooking, I weigh "significant" stuff like potatoes, carrots, rice, flour, but just go with the flow on stuff like onions, bell peppers and so on.

**Comment by gbosky** [comment]
Throwing in my 2 cents for what I think would be another possibility. I think expanding the nutrition info would also be a great QoL feature.

I have been on the search again for another nutrition app and recently ran into a self hosted one called [SparkyFitness](https://github.com/CodeWithCJ/SparkyFitness). While I think nutrition information may be on the fringe of Mealies purpose, I think it should support just enough information to allow a possible integration with SparkyFitness (or other nutrition apps). SparkyFitness does have a Mealie integration, however it just pulls the nutrition information on that recipe if provided. 

I think there could be some benefits to using a per ingredient as discussed above, but I think it would be more beneficial to have each app do what they are best at. My opinion is SparkyFitness could have all the nutrition data or conversions taken care of per ingredient if it knew the recipes ingredients, not just the nutrition info if its filled out on that recipe. I am not familiar with the Mealie API enough to know if ingredient info is already possible of being passed and SparkyFitness just didn't implement it that way. 

If that was a possibility, then maybe the next step would be possibly double integrate with SparkyFitness API and pass nutrition info back to a recipe in Mealie? SparkyFitness could then calculate the nutrition info per ingredient of the meal using its own food item database and pass that back to the recipe? It may be worth the time for building that functionality then building ingredient nutrition info/calcs into Mealie when others are doing it (and it seems involved how it does source and calculate info). 

I think this would rise the quality and ease of use for SparkyFitness, while also possibly improving Mealies nutritional information that can help me plan a balanced diet. At the end of the day I use Mealie to plan and cook, Sparkyfitness to track my nutrition/exercise and the overlap of cooking my own meals has a lot of potential (which I feel if you are focusing on your nutrition you are more likely to be making your own meals too so integrations for both would be huge QoL).

I also don't want to sound like SparkyFitness is the way togo, I am just using that as an example since its another self hosted solution that seems to be developed in a similar way with more focus on nutritional data and a growing community. As long as there is a possible solution for others to take on the nutrition info, I think Mealie could work well in having the API work with passing recipes/ingredients to those apps, then possibly have an integration with others for pulling info in for recipe nutrition.

**Reply by JurgenCruz** [reply]
I kind of prefer having everything in a single place. I don't like having to jump around different apps if possible

**Comment by wkbaran** [comment]
What about integrating the USDA database? https://fdc.nal.usda.gov/
It's available for download or API access rate limited to 1000 reqs/hr

**Comment by michael-genson** [comment]
FYI there is a PR in progress for this: https://github.com/mealie-recipes/mealie/pull/6747
